### ISSUE TITLE
Need the ability to query specific node OR disabling HWLB in runtime

### ISSUE THREAD

**Issue by roydahan** [scylladb/scylladb#9042]
At the moment when we need to query a specific key or would like to check a suspected data loss, there is no easy way to query from each specific node.
Currently, what we do  is "tracing ON" and run the same query many times hoping the HWLB eventually retrieve the result from all 3 replicas. (Doing it in parallel on few nodes).

This makes the debug process very hard and not straightforward.

Disabling HWLB not in runtime (requires a service restart) will change the state of the cache and the issue may disappear.

**Comment by roydahan** [comment]
@slivne please prioritize

**Comment by nyh** [comment]
Interesting request. Adding the ability to disable  HWLB at runtime of course makes sense. 

Beyond that, some dtests also want to check that data arrived to a specific node (e.g., that repair worked), and to achieve this these tests needed to kill the other nodes. we could have invented a consistency level `THIS_ONE` which is like `ONE` but the coordinator only tries itself as a replica (regardless of whether this is a good idea or even if this node has the requested data).

**Comment by eliransin** [comment]
@nyh I think this is a good idea, it might also help in the field.
The only surprise to me is that a node doesn't try itself if one is used and he is a valid replica, I thought this is the
current behavior.

**Comment by nyh** [comment]
> @nyh I think this is a good idea, it might also help in the field.
> The only surprise to me is that a node doesn't try itself if one is used and he is a valid replica, I thought this is the
> current behavior.

This is true (I think...) when HWLB is disabled, but not when HWLB is enabled: The purpose of HWLB is to **override** the client's decision of which coordinator to use: Normally a token-aware client knows which three replicas hold a requested item, and sends the request to one of the three at random. When sending the same request many times, the result is that each of the three replicas is used as the coordinator in 1/3rd of the requests. Now, when HWLB decides that one node should get only 1/5th (for example) of the requests instead of 1/3rd, it needs to forward some of the requests it got to other nodes - despite being able to answer these requests on its own.

By the way, I wonder if the same thing also happens in Cassandra with dynamic snitch - if a driver sends a request to one of the replicas at random, but this coordinator know that the other two replicas are faster, it may use them instead? (I don't know if this is true, I'm just raising a possibility - I couldn't find this documented anywhere).

**Comment by eliransin** [comment]
@nyh  For my own education:
Is HWLB takes place on the client software?
Here is my understanding of things:
1. A CQL request is sent to the coordinator by the client.
2. The app client chooses a connection (which also means coordinator) according to some internal strategy.
3. If the CL is ONE and the coordinator is only replica for the information then the coordinator queries it's own storage and returns the result.
4. Else, use HWLB to compose a set of replicas to send the request to.

Have I completely misunderstood or maybe just step 3  doesn't really exist and step 4 is executed unconditionally?

**Comment by nyh** [comment]
> @nyh For my own education:

Check out https://www.scylladb.com/2017/09/21/scylla-heat-weighted-load-balancing/ for a blog post we wrote when introducing HWLB, which explains how it works, and why.

> Is HWLB takes place on the client software?

**No**. The client is **not** aware of HWLB. What a token-aware client knows is which 3 nodes are replicas of the requested row, so instead of randomly picking any node - it randomly picks one of these 3 nodes, and HWLB works on the coordinator to "fix" this equal distribution by sending some of the requests that arrived to the slower node to the other nodes. The story is _signficantly_ more complicated when CL > 1, as explained in the aforementioned blog post.

> Here is my understanding of things:
> 
>     1. A CQL request is sent to the coordinator by the client.
> 
>     2. The app client chooses a connection (which also means coordinator) according to some internal strategy.

When we wrote HWLB, the assumption was (but @gleb-cloudius, please correct me if I'm wrong) that a token-aware client will pick on of the three replicas **at random**, and there is no other strategy.

Recently I heard someone suggests that maybe the choice isn't **random**, and in fact the client picks the connection to the replica which has the least number of unreplied requests. This doesn't necessarily result in 1/3rd of the requests being sent to each of the 3 replicas, as we assumed. If the 3 different coordinators have different _latencies_ for the requests, the lower latency coordinator will get more requests (the queue lengths are equal + lower latency results in higher throughput - this is Little's Law). I'm not sure yet if this is a real problem or the problem fixes itself (the coordinator with higher latencies will get fewer requests, so its latency will lower).

>     3. If the CL is ONE **AND** the coordinator is also a replica for the information then the coordinator queries it's own storage and returns the result.

No. This is explained in the blog post but in short: If HWLB decides that node X should only get half the requests per second as node Y or Z, and if each of X,Y,Z gets 1/3rd of all the requests - then X must send some of its requests to Y and Z - even at the cost of more network activity - otherwise X will get overloaded.

>     4. Else, use HWLB to compose a set of replicas to send the request to.
> 
> 
> Have I completely misunderstood or maybe just step 3 doesn't really exist and step 4 is executed unconditionally?

Step 3 doesn't exist. HOWEVER, what is indeed true is that when the cache is balanced (all three nodes have roughly the same cache hit), the result is that all three coordinators keep all of the requests they get and run them locally - without sending any of them to a different node. I explained in the blog post how this was a _goal_ of the HWLB algorithm - an algorithm where each node runs 1/3rd of the requests it gets and sends the rest to the two other nodes would have been equally correct, but much more wasteful in network bandwidth.

**Comment by eliransin** [comment]
@nyh thanks, the conclusion from the above is that I should read the blog post and maybe learn one or two things (or 1000 :smile: )

**Comment by nyh** [comment]
> It would be nice to have this for writes as well, although that would disqualify the use of `ANY` as a consistency level.

This would be a strange feature to tack on to the "consistency level", because Scylla's writes always write to **all** replicas, with the CL only specifying how many to wait for. Deliberately writing an update only to one replica and not to all of them might have some value in tests, but not for anything else as far as I can tell, which is why I cared more about reads. I think @roydahan who opened this issue also cared only about reads.

If we want to support "local-only" requests for both reads and write, maybe we can add some sort of CQL syntax (e.g., adding "ONLY LOCAL" at the end of the query) that will somehow cause our implementation of both reads and write to change what it does. Alternatively, we can do something like I think you did recently, wrapping the table's name in some keyword - for example if you SELECT FROM or INSERT INTO the virtual table `LOCAL_ONLY(mytab)` it will read or write from `mytab` with the extra trick that the coordinator will attempt to do this request locally if it can (we need to decide whether it will fail if the requested partition doesn't belong to it, or what).

> Another possibility is to open a separate CQL port, which serves all traffic strictly locally. Certain kind of queries (.e.g LWT) would be disallowed, because they are incompatible with the very nature of serving things from the local replica only.

Maybe, but it sounds to me like it's too ugly and complex to open yet another port - why is it better than adding some sort of option/flag/whatever on the query that signals that it is a "local" query? Also, I don't want this new feature configurable or turned off or firewalled, like ports often are - I want to be able to use it on a production machine (to do materialized view consistency checks).

**Comment by denesb** [comment]
>  Alternatively, we can do something like I think you did recently, wrapping the table's name in some keyword - for example if you SELECT FROM or INSERT INTO the virtual table LOCAL_ONLY(mytab) it will read or write from mytab with the extra trick that the coordinator will attempt to do this request locally if it can (we need to decide whether it will fail if the requested partition doesn't belong to it, or what).

This is a good idea and quite easy to do as well.

**Comment by roydahan** [comment]
thx, closing.
